### DISCUSSION TITLE
Realtime-api traces (model/vendor agnostic)

### DISCUSSION THREAD

**Discussion by marcklingen** [langfuse/langfuse#4757] [Ideas]
### Describe the feature or potential improvement

Requirements (current notes, let's collect more in thread)
-  capture streams
- for non-text provide timestamped transcript/labels. Unsure if Langfuse needs to provide support in generting these annotations or if they just need to be captured when models provide them in their output.
- capture the difference between user-time and api-level-time (e.g. when catching up on a stream after having categorized the input first).

### Additional information

https://github.com/orgs/langfuse/discussions/3765: Support for OpenAI Realtime API

**Comment by marcklingen** [comment]
This is something we have spent quite some time on already, both on designs internally and talking to teams using Langfuse with realtime applications. IMO the usage patterns are still developing and the data capturing/visualization needs across teams is not aligned.

**Comment by ghost** [comment]
Observability for realtime interactions would be super cool.  Subscribed for updates!

We too are using Langfuse already to capture some of our realtime interactions (currently with the OpenAI realtime API, but looking into Livekit), but our implementation is quite janky and only covers the bare minimum. Some efforts for the official team or rest of the community would be greatly appreciated.  
If needed I can make some time for a chat and show what our current setup looks like.

**Reply by marcklingen** [reply]
thanks for sharing! can you maybe share a loom video here? this might be interesting for everyone who is interested in this

**Reply by ghost** [reply]
I'm afraid a public Loom video won't be possible since we're at the risk of sharing sensitive data.

What I can share is this gist with some code snippets of our (very rudimentary and cobbled together) integration.  
https://gist.github.com/FlorDonnaSanders/8be72b06ecd2f82d20102877919ab74d

It is quite sketchy code, but if it can be of any help to someone building a more robust solution I am happy to share it.

**Reply by klentyboopathi** [reply]
@FlorDonnaSanders Thank you so much to make our job easy by providing the snippets :)

**Comment by peternixey** [comment]
Just observability on the text conversation would be excellent and provide a big win completely independently of audio. IMO The most useful end result of the OpenAI realtime API is the transcript itself rather than the audio element. 

I don't know what your API is for the assistant's API but FWIW just creating an adapter that plugged the realtime transcripts into that would go a very long way. 

WRT OpenAI, all the transcriptions happen asychronously so need a little work to get out. The OpenAI docs suck but the two events you want to watch for are `conversation.item.input_audio_transcription.completed` for the user's audio transcription and `response.done` for the assistant's. Those two may arrive out of order but the ordering comes from `conversation.items` which is the linked list that everything else gets hung off. 

There's about 37 events in total but to save you some days of diving these are the events you need to build the transcript:

- `conversation.item.create` - fired on the client when the user intiates a new message
- `conversation.item.created` - fired by the server when either a user's message or the assistant's message has been created
- `response.output_item.done` - fired when the assistant has finished streaming respones and includes the full text from the assistant
- `conversation.item.input_audio_transcription.completed` - fired by the server when the transcription of the user's message is complete. May fire after the next response from the assistant. Is also done (at the time of writing) by a separate model to the audio model so can be divergeant with what gets spoken. 

Because the OpenAI docs are all on a single page and you cant' feed them into an LLM to help you build with them I actually copied and pasted all of the pertinent events (not all of the events mind) into a page that I work with in Cursor. I've attached two files - one is that and another is  some zod schema and types for dealing with the events.
[realtime events + documentation.zip](https://github.com/user-attachments/files/18891148/realtime.events.%2B.documentation.zip)

**Reply by marcklingen** [reply]
thanks for sharing, super helpful context on this thread!

**Comment by abaybektursun** [comment]
So currently Langfuse does not support `gpt-4o-realtime-preview-*`? If so when can we expect a beta?
Is there partial support at least for things like token usage tracking?

**Reply by marcklingen** [reply]
Generally you can track any model in Langfuse via the low-level sdk or decorator. What is currently not supported is automated tracking via the wrapped openai sdk (the more native integration).

**Comment by nischalj10** [comment]
Any updates here as the realtime api has come out of beta?
